litr/src/main/java/com/linkedin/android/litr/io/ExternalMediaSource.kt

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright 2022 LinkedIn Corporation
+ * All Rights Reserved.
+ *
+ * Licensed under the BSD 2-Clause License (the "License").  See License in the project root for
+ * license information.
+ *
+ * Author: Ian Bird
+ */
+package com.linkedin.android.litr.io
+
+import android.media.MediaCodec
+import android.media.MediaFormat
+import android.util.Log
+import android.view.Surface
+import com.linkedin.android.litr.codec.Decoder
+import com.linkedin.android.litr.codec.Frame
+import com.linkedin.android.litr.codec.PassthroughDecoder
+import com.linkedin.android.litr.exception.MediaSourceException
+import java.nio.ByteBuffer
+
+private const val TAG = "ExternalMediaSource"
+private const val DECODER_NAME = "ExternalMediaSource.Decoder"
+
+private const val MIME_TYPE_RAW = "video/raw"
+private const val DEFAULT_FRAME_RATE = 30
+private const val DEFAULT_BITRATE = 10_000_000
+private const val DEFAULT_KEY_FRAME_INTERVAL = 5
+
+/**
+ * An implementation of MediaSource, which exposes the input {@link Surface} of the encoder via a
+ * Callback. An instance of this class is expected to be both the MediaSource, and Decoder for which
+ * the pipeline is configured. This allows these components to be bypassed.
+ */
+class ExternalMediaSource(private val callback: Callback): MediaSource, Decoder {
```

Comment:
While looking into how to implement `Camera2MediaSource`, it shows that requiring the callback via the constructor isn't that flexible. It's arguably better to change to something more like `addCallback` (or `addListener`). That's a lot more flexible.

Happy to make the change as part of this PR, or the follow up where I utilise it properly in the `Camera2MediaSource`